### PR TITLE
#2454 Handle the parsing of json arrays to handle duplicate array values

### PR DESCRIPTION
**Pull request by brianpos**
(both orders and extension/value)

## Description
Update to json parser to handle the case.

## Related issues
Fixes issue #2454 

## Testing
Unit test included to verify functionality

## FirelyTeam Checklist
- [ ] **Update the title** of the PR to be succinct and less than 50 characters
- [ ] Mark the PR with the label **breaking change** when this PR introduces breaking changes

### CODE REVIEW

File: src/Hl7.Fhir.Base/Serialization/BaseFhirJsonPocoDeserializer.cs [hunk 1]
```diff
@@ -510,6 +524,12 @@ FhirJsonPocoDeserializerState state
             // to simply create a list by Adding(). Not the fastest approach :-(
             int elementIndex = 0;
             bool? onlyNulls = null;
+            elementIndex = delayedValidations.ParsingPropertyValue(propertyName);
+            if (elementIndex > 0)
+            {
+                state.Path.IncrementIndex(elementIndex);
+                state.Errors.Add(ERR.DUPLICATE_ARRAY(ref reader, state.Path.GetInstancePath()));
```
Comment (ewoutkramer) [inline src/Hl7.Fhir.Base/Serialization/BaseFhirJsonPocoDeserializer.cs:531]: Add duplicate array to the list of recoverable errors.

File: src/Hl7.Fhir.Base/Serialization/BaseFhirJsonPocoDeserializer.cs [hunk 2]
```diff
@@ -482,7 +496,7 @@ public void Run()
             ClassMapping propertyValueMapping,
             Type? fhirType,
             ref Utf8JsonReader reader,
-            DelayedValidations delayedValidations,
+            ObjectParsingState delayedValidations,
```
Comment (ewoutkramer) [inline src/Hl7.Fhir.Base/Serialization/BaseFhirJsonPocoDeserializer.cs:499]: Wouldn't another name for the variable better express the intentions now?

File: src/Hl7.Fhir.Base/Serialization/BaseFhirJsonPocoDeserializer.cs [hunk 3]
```diff
@@ -454,19 +454,33 @@ FhirJsonPocoDeserializerState state
...
-            public void Schedule(string key, Action validation)
+            public int ParsingPropertyValue(string memberName)
+            {
+                if (_parsedPropValue.ContainsKey(memberName))
+                    return _parsedPropValue[memberName];
+                _parsedPropValue.Add(memberName, 0);
+                return 0;
+            }
+
+            public void ParsedPropertyValue(string memberName, int count)
```
Comment (ewoutkramer) [inline src/Hl7.Fhir.Base/Serialization/BaseFhirJsonPocoDeserializer.cs:470]: SetPropertyIndex?

File: src/Hl7.Fhir.Base/Serialization/BaseFhirJsonPocoDeserializer.cs [hunk 4]
```diff
@@ -454,19 +454,33 @@ FhirJsonPocoDeserializerState state
...
         }
 
-        internal class DelayedValidations
+        internal class ObjectParsingState
         {
             private readonly Dictionary<string, Action> _validations = new();
+            private readonly Dictionary<string, int> _parsedPropValue = new();
 
-            public void Schedule(string key, Action validation)
+            public int ParsingPropertyValue(string memberName)
```
Comment (ewoutkramer) [inline src/Hl7.Fhir.Base/Serialization/BaseFhirJsonPocoDeserializer.cs:462]: GetPropertyIndex?

**Review by ewoutkramer — CHANGES_REQUESTED** [review]
Minor things...

**Comment by mmsmits** [comment]
fixes #2465

File: src/Hl7.Fhir.Base/Serialization/BaseFhirJsonPocoDeserializer.cs (hunk 1 above)
Comment (mmsmits) [reply src/Hl7.Fhir.Base/Serialization/BaseFhirJsonPocoDeserializer.cs:531]: I think this is already added

**Review by ewoutkramer — APPROVED** [review]

**Pull request merged**